File: checktick_app/surveys/branching.py
"""Survey branching logic evaluation."""

from typing import Any

from .models import SurveyQuestion, SurveyQuestionCondition


def evaluate_condition(condition: SurveyQuestionCondition, answer: Any) -> bool:
    """
    Evaluate whether a condition is met based on the user's answer.

    Args:
        condition: The condition to evaluate
        answer: The user's answer to the question

    Returns:
        True if the condition is met, False otherwise
    """
    # Handle empty/missing answers
    if answer is None or answer == "" or answer == []:
        has_answer = False
    else:
        has_answer = True

    # Operators that don't require a value
    if condition.operator == SurveyQuestionCondition.Operator.EXISTS:
        return has_answer
    elif condition.operator == SurveyQuestionCondition.Operator.NOT_EXISTS:
        return not has_answer

    # All other operators require an answer to compare against
    if not has_answer:
        return False

    # Convert answer to string for comparison
    # Handle multiple choice (list) by joining
    if isinstance(answer, list):
        answer_str = ",".join(str(a) for a in answer)
    else:
        answer_str = str(answer)

    value = condition.value or ""

    # String comparisons (case-insensitive)
    if condition.operator == SurveyQuestionCondition.Operator.EQUALS:
        return answer_str.lower().strip() == value.lower().strip()
    elif condition.operator == SurveyQuestionCondition.Operator.NOT_EQUALS:
        return answer_str.lower().strip() != value.lower().strip()
    elif condition.operator == SurveyQuestionCondition.Operator.CONTAINS:
        return value.lower() in answer_str.lower()
    elif condition.operator == SurveyQuestionCondition.Operator.NOT_CONTAINS:
        return value.lower() not in answer_str.lower()

    # Numeric comparisons
    try:
        answer_num = float(answer_str)
        value_num = float(value)

        if condition.operator == SurveyQuestionCondition.Operator.GREATER_THAN:
            return answer_num > value_num
        elif condition.operator == SurveyQuestionCondition.Operator.GREATER_EQUAL:
            return answer_num >= value_num
        elif condition.operator == SurveyQuestionCondition.Operator.LESS_THAN:
            return answer_num < value_num
        elif condition.operator == SurveyQuestionCondition.Operator.LESS_EQUAL:
            return answer_num <= value_num
    except (ValueError, TypeError):
        # If conversion fails, comparison fails
        return False

    return False


def get_visible_questions(
    all_questions: list[SurveyQuestion], answers: dict[str, Any]
) -> tuple[list[SurveyQuestion], bool]:
    """
    Determine which questions should be visible based on branching logic.

    Args:
        all_questions: All questions in the survey (in order)
        answers: Dictionary mapping question IDs to answers

    Returns:
        Tuple of (visible_questions, survey_ended)
        - visible_questions: List of questions that should be shown
        - survey_ended: True if END_SURVEY condition was triggered
    """
    visible = []
    survey_ended = False
    skip_until_idx = None  # Used for JUMP_TO logic

    for idx, question in enumerate(all_questions):
        question_id = str(question.id)

        # If we're in a skip mode (from JUMP_TO), check if we've reached the target
        if skip_until_idx is not None:
            if idx < skip_until_idx:
                continue  # Skip this question
            else:
                skip_until_idx = None  # Reached target, resume normal flow

        # Check if this question has been answered
        answer = answers.get(question_id)

        # Load conditions for this question (should be prefetched)
        try:
            conditions = list(question.conditions.all().order_by("order"))
        except Exception:
            conditions = []

        # Evaluate conditions
        triggered_action = None
        triggered_target = None

        for condition in conditions:
            if evaluate_condition(condition, answer):
                triggered_action = condition.action
                triggered_target = condition.target_question

                if condition.action == SurveyQuestionCondition.Action.END_SURVEY:
                    survey_ended = True
                    break  # Stop processing this question's conditions
                elif condition.action == SurveyQuestionCondition.Action.JUMP_TO:
                    # Find the index of the target question
                    if condition.target_question:
                        try:
                            target_idx = next(
                                i
                                for i, q in enumerate(all_questions)
                                if q.id == condition.target_question.id
                            )
                            skip_until_idx = target_idx
                        except StopIteration:
                            pass
                    break  # First matching condition wins
                elif condition.action == SurveyQuestionCondition.Action.SKIP:
                    # Skip the target question (don't show it)
                    # This is handled by not adding the target to visible
                    break
                elif condition.action == SurveyQuestionCondition.Action.SHOW:
                    # Show the target question
                    # This will be handled when we reach that question
                    break

        # If survey ended, don't show any more questions
        if survey_ended:
            break

        # Add this question to visible list
        visible.append(question)

        # If SKIP was triggered, find and skip the target question
        if triggered_action == SurveyQuestionCondition.Action.SKIP and triggered_target:
            # Mark the target question to be skipped
            # We'll filter it out by not adding it when we reach it
            pass

    return visible, survey_ended


def should_show_question(
    question: SurveyQuestion,
    all_questions: list[SurveyQuestion],
    answers: dict[str, Any],
) -> bool:
    """
    Determine if a specific question should be shown based on branching logic.

    This handles SHOW conditions - a question with incoming SHOW conditions
    should only be visible if at least one of those conditions is met.

    Args:
        question: The question to check
        all_questions: All questions in the survey
        answers: Current answers

    Returns:
        True if the question should be shown
    """
    # Check if there are any incoming SHOW conditions
    try:
        show_conditions = SurveyQuestionCondition.objects.filter(
            target_question=question, action=SurveyQuestionCondition.Action.SHOW
        )

        if not show_conditions.exists():
            # No SHOW conditions, question is always visible
            return True

        # Check if any SHOW condition is met
        for condition in show_conditions:
            source_question_id = str(condition.question.id)
            answer = answers.get(source_question_id)
            if evaluate_condition(condition, answer):
                return True

        # No SHOW conditions were met
        return False
    except Exception:
        # If there's an error, default to showing the question
        return True

File: checktick_app/static/js/branching.js
/**
 * Survey branching logic - dynamically show/hide questions based on conditions
 */

(function () {
  "use strict";

  // Get branching configuration from data attribute
  const form = document.querySelector("[data-survey-form]");
  if (!form) return;

  const branchingData = form.dataset.branchingConfig;
  if (!branchingData) return;

  let config;
  try {
    config = JSON.parse(branchingData);
  } catch (e) {
    console.error("Failed to parse branching config:", e);
    return;
  }

  /**
   * Evaluate a single condition based on current answer
   */
  function evaluateCondition(condition, answer) {
    // Handle empty/missing answers
    const hasAnswer =
      answer !== null &&
      answer !== undefined &&
      answer !== "" &&
      !(Array.isArray(answer) && answer.length === 0);

    // Operators that don't require a value
    if (condition.operator === "exists") {
      return hasAnswer;
    } else if (condition.operator === "not_exists") {
      return !hasAnswer;
    }

    // All other operators require an answer
    if (!hasAnswer) return false;

    // Convert answer to string for comparison
    let answerStr;
    if (Array.isArray(answer)) {
      answerStr = answer.join(",");
    } else {
      answerStr = String(answer);
    }

    const value = condition.value || "";

    // String comparisons (case-insensitive)
    const answerLower = answerStr.toLowerCase().trim();
    const valueLower = value.toLowerCase().trim();

    switch (condition.operator) {
      case "eq":
        return answerLower === valueLower;
      case "neq":
        return answerLower !== valueLower;
      case "contains":
        return answerLower.includes(valueLower);
      case "not_contains":
        return !answerLower.includes(valueLower);

      // Numeric comparisons
      case "gt":
      case "gte":
      case "lt":
      case "lte":
        try {
          const answerNum = parseFloat(answerStr);
          const valueNum = parseFloat(value);
          if (isNaN(answerNum) || isNaN(valueNum)) return false;

          switch (condition.operator) {
            case "gt":
              return answerNum > valueNum;
            case "gte":
              return answerNum >= valueNum;
            case "lt":
              return answerNum < valueNum;
            case "lte":
              return answerNum <= valueNum;
          }
        } catch (e) {
          return false;
        }
        break;
    }

    return false;
  }

  /**
   * Get current answer for a question
   */
  function getAnswer(questionId) {
    const inputs = form.querySelectorAll(`[name="q_${questionId}"]`);
    if (inputs.length === 0) return null;

    const firstInput = inputs[0];

    if (firstInput.type === "checkbox") {
      // Multiple choice - collect all checked values
      const checked = Array.from(inputs)
        .filter((inp) => inp.checked)
        .map((inp) => inp.value);
      return checked.length > 0 ? checked : null;
    } else if (firstInput.type === "radio") {
      // Single choice - find checked radio
      const checked = Array.from(inputs).find((inp) => inp.checked);
      return checked ? checked.value : null;
    } else {
      // Text, number, select, etc.
      return firstInput.value || null;
    }
  }

  /**
   * Update visibility of all questions based on current answers
   */
  function updateVisibility() {
    const answers = {};

    // Collect all answers
    config.questions.forEach((qid) => {
      answers[qid] = getAnswer(qid);
    });

    let surveyEnded = false;
    const skipQuestions = new Set();

    // Process each question's conditions
    config.questions.forEach((questionId) => {
      const questionConditions = config.conditions[questionId] || [];
      const answer = answers[questionId];

      // Evaluate conditions for this question
      for (const condition of questionConditions) {
        if (evaluateCondition(condition, answer)) {
          // Condition is met - apply action
          switch (condition.action) {
            case "end_survey":
              surveyEnded = true;
              break;
            case "skip":
              if (condition.target_question) {
                skipQuestions.add(condition.target_question);
              }
              break;
            case "jump_to":
              // Hide all questions between current and target
              if (condition.target_question) {
                const currentIdx = config.questions.indexOf(questionId);
                const targetIdx = config.questions.indexOf(
                  condition.target_question
                );
                if (currentIdx !== -1 && targetIdx !== -1) {
                  for (let i = currentIdx + 1; i < targetIdx; i++) {
                    skipQuestions.add(config.questions[i]);
                  }
                }
              }
              break;
          }
          // First matching condition wins
          break;
        }
      }
    });

    // Apply visibility changes
    config.questions.forEach((questionId, idx) => {
      const questionElement = form.querySelector(
        `[data-question-id="${questionId}"]`
      );
      if (!questionElement) return;

      // Hide if survey ended and this is after the question that ended it
      const currentQuestionAnswer = answers[questionId];
      const currentConditions = config.conditions[questionId] || [];
      const thisQuestionEnded = currentConditions.some(
        (c) =>
          c.action === "end_survey" &&
          evaluateCondition(c, currentQuestionAnswer)
      );

      if (surveyEnded && !thisQuestionEnded) {
        // Find which question triggered the end
        for (let i = 0; i < idx; i++) {
          const prevQid = config.questions[i];
          const prevAnswer = answers[prevQid];
          const prevConditions = config.conditions[prevQid] || [];
          const prevEnded = prevConditions.some(
            (c) => c.action === "end_survey" && evaluateCondition(c, prevAnswer)
          );
          if (prevEnded) {
            questionElement.style.display = "none";
            return;
          }
        }
      }

      // Hide if this question should be skipped
      if (skipQuestions.has(questionId)) {
        questionElement.style.display = "none";
        return;
      }

      // Handle SHOW conditions
      const showConditions = config.show_conditions[questionId] || [];
      if (showConditions.length > 0) {
        // This question has SHOW conditions - only show if one is met
        let shouldShow = false;
        for (const condition of showConditions) {
          const sourceAnswer = answers[condition.source_question];
          if (evaluateCondition(condition, sourceAnswer)) {
            shouldShow = true;
            break;
          }
        }
        questionElement.style.display = shouldShow ? "" : "none";
        return;
      }

      // Default: show the question
      questionElement.style.display = "";
    });
  }

  // Attach listeners to all form inputs
  form.addEventListener("input", updateVisibility);
  form.addEventListener("change", updateVisibility);

  // Run initial visibility check
  updateVisibility();
})();
